CardanoSharp.Wallet/Models/Addresses/Address.cs

```diff
@@ -0,0 +1,131 @@
+﻿using CardanoSharp.Wallet.Encoding;
+using CardanoSharp.Wallet.Enums;
+using System;
+using System.Collections.Generic;
+using System.Text;
+
+namespace CardanoSharp.Wallet.Models.Addresses
+{
+    public class Address : IEqualityComparer<Address>, IEquatable<Address>, IEquatable<string>, IEquatable<byte[]>
+    {
+        private byte[] _bytes;
+        private string _address;
+        public AddressType AddressType { get; }
+        public NetworkType NetworkType { get; }
+        
+        public Address()
+        {
+        }
+
+        public Address(string prefix, byte[] address)
+        {
+            var bech32 = new Bech32();
+            _bytes = address;
+            _address = bech32.Encode(address, prefix);
+        }
+
+        public Address(string address)
+        {
+            if (string.IsNullOrWhiteSpace(address)) 
+                throw new ArgumentNullException(nameof(address));
+
+            var bech32 = new Bech32();
+            
+            //if (!bech32.HasValidChars(address))
+            //{
+            //    throw new ArgumentException("Invalid characters", nameof(address));
+            //}
+
+            _address = address;
+            _bytes = bech32.Decode(_address, out byte witVer, out string prefix);
+
+            Prefix = prefix;
+            WitnessVersion = witVer;
+            AddressType = GetAddressType();
+            NetworkType = GetNetworkType();
+        }
+
+        /// <summary>
+        /// Returns <see cref="AddressType"/> as defined in https://github.com/cardano-foundation/CIPs/blob/master/CIP-0019/CIP-0019.md
+        /// </summary>
+        /// <returns>
+        ///     <para><see cref="AddressType.Base"/> if AddressType header is 0x01</para>
+        ///     <para><see cref="AddressType.Enterprise"/> if AddressType header is 0x06</para>
+        ///     <para>otherwise <see cref="AddressType.Base"/></para>
+        /// </returns>
+        /// <returns></returns>
+        private AddressType GetAddressType()
+        {
+            return (_bytes[0] >> 4) switch
+            {
+                0x01 => AddressType.Base, //@TODO: derive all AddressTypes
+                0x06 => AddressType.Enterprise,
```

Comment:
I believe Reward type is the same as Enterprise. I haven't done a lot of testing around this, but thats what i came up with from reading the Rust code

Reply:
after reviewing the `AddressService`, i believe im wrong